### ISSUE TITLE
Manual Mediator Registration Issue (v3.0.1->v5.0.1)

### ISSUE THREAD

**Issue by erdmenchen** [jbogard/MediatR#296]
First: Thank you very much for this awesome project, it has inspired me a lot and structures my project architecture so much better. Thanks!

I am currently migrating from the old version 3.0.1 to the latest 5.0.1.
Since I use the `SimpleIoc` from the MvvmLight framework, this was how I registered and setup MediatR in version 3:

```
var ioc = SimpleIoc.Default;
ioc.Register<IAsyncRequestHandler<DoJobRequest>, DoJobUseCase>();

var mediator = new Mediator(
          new SingleInstanceFactory(ioc.GetInstance),
          new MultiInstanceFactory(ioc.GetAllInstances)
);

await mediator.Send(new DoJobRequest());
```

With version 5.0.1 and the removal of the `MultiInstanceFactory`, I tried this (as described in the migration wiki):

```
var ioc = SimpleIoc.Default;
ioc.Register<IRequestHandler<DoJobRequest>, DoJobUseCase>();

var mediator = new Mediator(
          new ServiceFactory(ioc.GetInstance)
);

await mediator.Send(new DoJobRequest());
```

But I get the error, that the `IEnumerable<IPipelineBehavior<...>>` was not found...
How should I setup MediatR now to get it going again?

Thank you very much.

**Comment by jbogard** [comment]
What is SimpleIoc? I don't have that in the list of samples I keep track
of. Do you have a link?

**Comment by erdmenchen** [comment]
See here:
https://github.com/lbugnion/mvvmlight/blob/master/GalaSoft.MvvmLight/GalaSoft.MvvmLight.Extras%20(PCL)/Ioc/SimpleIoc.cs

It implements the `IServiceLocator `interface from the `CommonServiceLocator`: https://github.com/unitycontainer/commonservicelocator

Its the dependency container from the very popular MVVMLight Toolkit: https://github.com/lbugnion/mvvmlight

**Comment by jbogard** [comment]
Ah, so this might be one of the containers that can't handle resolving `IEnumerable<T>` without calling the special overload. Some of the older containers are like this. Here's what I did for Castle.Windsor:

https://github.com/jbogard/MediatR/blob/master/samples/MediatR.Examples.Windsor/Program.cs#L35

You'll likely need to do something similar - check to see if it's `IEnumerable` and take appropriate action.

**Comment by erdmenchen** [comment]
I tried it but it throws an InvalidCast-Exception:
It cannot convert `System.Collections.Generic.List`1[System.Object]` to `System.Collections.Generic.IEnumerable`1[MediatR.IPipelineBehavior`2[DoJobRequest,MediatR.Unit]]`.

This is my current ServiceFactory-Delegate copied from your Windsor example:
```
var enumerableType = type
                    .GetInterfaces()
                    .Concat(new[] { type })
                    .FirstOrDefault(t => t.IsGenericType && t.GetGenericTypeDefinition() == typeof(IEnumerable<>));

if (enumerableType != null)
{
    var instances = _ioc.GetAllInstances(enumerableType.GetGenericArguments()[0]).ToList();
    return instances;
}
else
{
    var instance = _ioc.GetInstance(type);
    return instance;
}

```

Is there no example for the general IServiceLocator-Implementation?

**Comment by jbogard** [comment]
Nope, multiple overloads is much, much worse for DI containers. And the feature change is why the major version bumped. I made this change to simplify things, knowing some older/less feature-complete containers would need special handling code. Most containers implicitly handle enumerable dependencies:

- MS.Extensions.DI - yes
- AutoFac - yes
- DryIoc - yes
- DryIocZero - yes
- Lamar - yes
- LightInject - yes
- Ninject - yes
- SimpleInjector - yes
- StructureMap - yes
- Unity - no
- Windsor - no

So basically all the new, modern, active DI containers and the more less used, much older, much less maintained containers do not. So rather than make things harder for the 99%, I thought to make it harder for the 1%.

I suspect your code is wrong, however. You're calling `ToList` to convert a runtime `IEnumerable<T>` to a `List<object>`. Remove that `ToList` and I bet it will work just fine.

**Comment by jbogard** [comment]
If you get yours working btw, I'm more than happy to add it to the samples for future folks to see. Then I can add it to the feature support matrix:

https://github.com/jbogard/MediatR/wiki/Container-Feature-Support

**Comment by erdmenchen** [comment]
Thank you very much for your detailed answer. I appreciate it.

> I suspect your code is wrong, however. You're calling ToList to convert a runtime IEnumerable<T> to a List<object>. Remove that ToList and I bet it will work just fine.

`SimpleIoC.GetAllInstances()` returns a `ValueCollection` from a dictionary and not a list.
Changing this does not fix it.

So I am basically stuck with MediatR v3, because I cannot switch the DI container unless there is a container that implements the IServiceLocator interface.

I think there is not much missing to get this running, I just don´t know how to convert the List<object> to the requested IEnumerable<...>.

**Comment by Brimce** [comment]
Hi,
I'm in the same situation with [splat](https://github.com/reactiveui/splat) container.

My solution :
```csharp
 _dependencyResolver.RegisterLazySingleton<ServiceFactory>(() => (type =>
            {
                if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(IEnumerable<>))
                {
                    var enumerableType = type.GetGenericArguments()[0];

                    var pipelineInfo = enumerableType.GetGenericArguments()[0];

                    var requestType = pipelineInfo
                        .GetInterfaces()
                        .FirstOrDefault(type1 => type1.IsGenericType && type1.GetGenericTypeDefinition() == typeof(IRequest<>));

                    var responseType = requestType.GetGenericArguments()[0];

                    var typeArguments = new Type[2];
                    typeArguments[0] = requestType;
                    typeArguments[1] = responseType;
                    var pipelineType = typeof(GenericPipelineBehavior<,>).MakeGenericType(typeArguments);

                    var enumerableTypeResult = typeof(List<>).MakeGenericType(pipelineType);
                    var list = (IList) Activator.CreateInstance(enumerableTypeResult);

                    var pipeline = Activator.CreateInstance(typeof(GenericPipelineBehavior<,>).MakeGenericType(requestType, responseType));

                    list.Add(pipeline);

                    return list;
                }

                return _dependencyResolver.GetService(type);
            }));
```
Just replace _dependencyResolver with _ioc and add the folowing class:

```csharp
 public class GenericPipelineBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse>
 {        
        public async Task<TResponse> Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<TResponse> next)
        {
            //log... "-- Handling Request"
            var response = await next();
            //log "-- Finished Request"
            return response;
        }
}
```
I Hope it can help !

**Comment by erdmenchen** [comment]
Awesome @Brimce ! This was the right direction hint, I needed.

Since I do have multiple Instances registered for some requests, I needed the real instances from the IoC.
But with your conversion of setting up the (empty) list of results, I was able to put it all together like this:

```
        object Resolve(Type type)
        {
            var enumerableType = type
                    .GetInterfaces()
                    .Concat(new[] { type })
                    .FirstOrDefault(t => t.IsGenericType && t.GetGenericTypeDefinition() == typeof(IEnumerable<>));

            if (enumerableType != null)
            {
                var targetType = type.GetGenericArguments()[0];
                var responseType = typeof(Unit);
                var requestType = targetType.GetGenericArguments()[0];
                var requestTypeArguments = requestType.GetGenericArguments();
                if (!requestTypeArguments.Any())
                {
                    requestTypeArguments = requestType
                        .GetInterfaces()
                        .FirstOrDefault(t => t.IsGenericType && t.GetGenericTypeDefinition() == typeof(IRequest<>))
                        .GetGenericArguments();
                }
                if (requestTypeArguments.Any())
                    responseType = requestTypeArguments[0];

                var typeArguments = new Type[] { requestType, responseType };
                var pipelineType = typeof(IPipelineBehavior<,>).MakeGenericType(typeArguments);

                var enumerableTypeResult = typeof(List<>).MakeGenericType(pipelineType);
                var list = (IList)Activator.CreateInstance(enumerableTypeResult);

                var instances = _ioc.GetAllInstances(targetType);
                if (instances.Any())
                    instances.ToList().ForEach(x => list.Add(x));

                return list;
            }

            return _ioc.GetInstance(type);
        }

```

Also I eliminated the need of the `GenericPipelineBehavior` as I replaced the list type with `IPipelineBehavior<,>`

I still need to make some tests with all various kinds of requests and also with real pipeline requests, but I think, this might be the solution...

**Comment by jbogard** [comment]
This all leads me to think I might need some sort of built-in way of doing
all this junk. This looks ugly.

**Comment by erdmenchen** [comment]
Yea so I created this static helper class for the resolve delegate:

```
public static class ServiceResolver
    {
        public static object Resolve(
                Type type,
                Func<Type, object> singleInstance,
                Func<Type, IEnumerable<object>> multiInstance)
        {
            var enumerableType = type
                    .GetInterfaces()
                    .Concat(new[] { type })
                    .FirstOrDefault(t => t.IsGenericType && t.GetGenericTypeDefinition() == typeof(IEnumerable<>));

            if (enumerableType != null)
            {
                var targetType = type.GetGenericArguments()[0];
                var responseType = typeof(Unit);
                var requestType = targetType.GetGenericArguments()[0];
                var requestTypeArguments = requestType.GetGenericArguments();
                if (!requestTypeArguments.Any())
                {
                    requestTypeArguments = requestType
                        .GetInterfaces()
                        .FirstOrDefault(t => t.IsGenericType && t.GetGenericTypeDefinition() == typeof(IRequest<>))
                        .GetGenericArguments();
                }
                if (requestTypeArguments.Any())
                    responseType = requestTypeArguments[0];

                var typeArguments = new Type[] { requestType, responseType };
                var pipelineType = typeof(IPipelineBehavior<,>).MakeGenericType(typeArguments);

                var enumerableTypeResult = typeof(List<>).MakeGenericType(pipelineType);
                var list = (IList)Activator.CreateInstance(enumerableTypeResult);

                var instances = multiInstance(targetType);
                if (instances.Any())
                    instances.ToList().ForEach(x => list.Add(x));

                return list;
            }

            return singleInstance(type);
        }
    }
```
And registering works like this (notice that I had to add `Unit` to the request generics):

```
var ioc = SimpleIoc.Default;      // from MvvmLight Toolkit
ioc.Register<IRequestHandler<DoJobRequest, Unit>, DoJobUseCase>();

var mediator = new Mediator(
                new ServiceFactory(t => ServiceResolver.Resolve(t, ioc.GetInstance, ioc.GetAllInstances))
);

await mediator.Send(new DoJobRequest());

```


This seems to work with basic `IRequestHandler` registrations, as well as with pipeline behaviors like `RequestPreProcessorBehavior` or `RequestPostProcessorBehavior`.


Example for pipeline behavior:
```
ioc.Register<DoJobUseCase>();
ioc.Register<IRequestPreProcessor<DoJobRequest>>(() => ioc.GetInstance<DoJobUseCase>());
ioc.Register(ioc.GetAllInstances<IRequestPreProcessor<DoJobRequest>>);
ioc.Register<IPipelineBehavior<DoJobRequest, Unit>, RequestPreProcessorBehavior<DoJobRequest, Unit>>();
```

**Comment by erdmenchen** [comment]
Maybe @jbogard you want to include the static helper method in the library?
This is basically the lost implementation of split registration of single and multi instance delegate...

All what it is doing is the conversion between `List<object>` and `IEnumerable<IRequest<>`...

**Comment by jbogard** [comment]
Proooooobably not something like this. I'm thinking more of the scenario of "I will wrap your ugly implementation with my own delegate" or similar.

**Comment by erdmenchen** [comment]
Ok I am looking forward to it! Thank you very much for the support until now. This is awesome.

Alternatively you could also somehow make a ServiceFactory that just wraps an `IServiceLocator` from the `CommonServiceLocator` (as stated above). But that is currently not possible because the ServiceFactory has extension methods, that cannot be overriden...

**Comment by jbogard** [comment]
The problem that `IServiceLocator` is pretty much a dead end. A very old version of MediatR used `IServiceLocator`, but once that concept got abandoned, as did my usage of it.

Interestingly, the new ASP.NET Core DI is just `IServiceProvider`, which has existed in .NET since...forever. Everything else is extension methods on that, or extensions like [`IServiceScope`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.iservicescope?view=aspnetcore-2.1)
